modules/standard/CTypes.chpl

```diff
@@ -995,6 +995,79 @@ module CTypes {
     compilerError("Cannot call c_offsetof on type that is not a record");
   }
 
+  /* Allocate memory.
+
+     This uses the Chapel allocator. Memory allocated with this function should
+     eventually be freed with :proc:`deallocate`.
+
+    :arg eltType: the type of the elements to allocate
+    :arg size: the number of elements to allocate space for
+    :arg clear: whether to initialize all bits of allocated memory to 0
+    :arg alignment: Memory alignment of the allocation, which must be a power of
+                    two and a multiple of `c_sizeof(c_void_ptr)`. Alignment of 0
+                    is invalid and taken to mean default alignment.
+    :returns: a c_ptr(eltType) to allocated memory
+   */
+  @unstable("'allocate' is unstable, and may be renamed or moved")
+  inline proc allocate(type eltType, size: c_size_t, clear: bool = false,
+      alignment: c_size_t = 0) : c_ptr(eltType) {
+    const alloc_size = size * c_sizeof(eltType);
+    const aligned : bool = (alignment != 0);
+    var ptr : c_void_ptr = nil;
+
+    // pick runtime allocation function based on requested zeroing + alignment
+    if (!aligned) {
+      if (clear) {
+        // normal calloc
+        ptr = chpl_here_calloc(alloc_size, 1, offset_ARRAY_ELEMENTS);
+      } else {
+        // normal malloc
+        ptr = chpl_here_alloc(alloc_size, offset_ARRAY_ELEMENTS);
+      }
+    } else {
+      // check alignment, size restriction
+      // Alignment of 0 is our sentinel value for no specified alignment,
+      // so no need to check for it.
+      if boundsChecking {
+        use Math;
+        var one:c_size_t = 1;
+        // Round the alignment up to the nearest power of 2
+        var p = log2(alignment); // power of 2 rounded down
+        // compute alignment rounded up
+        if (one << p) < alignment then
+          p += 1;
+        assert(alignment <= (one << p));
+        if alignment != (one << p) then
+          halt("allocate called with non-power-of-2 alignment ", alignment);
+        if alignment < c_sizeof(c_void_ptr) then
+          halt("allocate called with alignment smaller than pointer size");
+      }
+
+      if (clear) {
+        // there is no aligned calloc; have to aligned_alloc + memset to 0
+        ptr = chpl_here_aligned_alloc(alignment, alloc_size,
+                                       offset_ARRAY_ELEMENTS);
+        c_memset(ptr, 0, alloc_size);
+      } else {
+        // normal aligned_alloc
+        ptr = chpl_here_aligned_alloc(alignment, alloc_size,
+                                       offset_ARRAY_ELEMENTS);
+      }
```

Comment:
Since the `chpl_here_aligned_alloc` calls are the same, can you reduce the repetition of code for this?

```chapel
ptr = ...;
if clear then c_memset(...)
```

 I think it is fine as is, but this is a little cleaner to my mind.

Reply:
Oh, good point, will definitely do that.